samples/bluetooth/alexa_gadget/src/gadgets/CMakeLists.txt

```diff
@@ -0,0 +1,181 @@
+    ${ALEXA_EMBEDDED_PATH}/AlexaGadgetsProtobuf/Alerts/DeleteAlert/alertsDeleteAlertDirectivePayload.proto
+    ${ALEXA_EMBEDDED_PATH}/AlexaGadgetsProtobuf/Alexa.Gadget.StateListener/StateUpdate/alexaGadgetStateListenerStateUpdateDirective.proto
+    ${ALEXA_EMBEDDED_PATH}/AlexaGadgetsProtobuf/Alexa.Gadget.StateListener/StateUpdate/alexaGadgetStateListenerStateUpdateDirectivePayload.proto
+    ${ALEXA_EMBEDDED_PATH}/AlexaGadgetsProtobuf/Alexa.Gadget.SpeechData/Speechmarks/alexaGadgetSpeechDataSpeechmarksDirective.proto
+    ${ALEXA_EMBEDDED_PATH}/AlexaGadgetsProtobuf/Alexa.Gadget.SpeechData/Speechmarks/alexaGadgetSpeechDataSpeechmarksDirectivePayload.proto
+    ${ALEXA_EMBEDDED_PATH}/AlexaGadgetsProtobuf/Alexa.Gadget.MusicData/Tempo/alexaGadgetMusicDataTempoDirective.proto
+    ${ALEXA_EMBEDDED_PATH}/AlexaGadgetsProtobuf/Alexa.Gadget.MusicData/Tempo/alexaGadgetMusicDataTempoDirectivePayload.proto
+    ${CMAKE_CURRENT_SOURCE_DIR}/proto/custom_event.proto
+)
+
+# Populate list of expected .c files output from protobuf generation
+foreach(file ${PROTO_FILES})
+    get_filename_component(FILE_PATH ${file} DIRECTORY)
+    get_filename_component(FILE_BASE ${file} NAME_WE)
+    list(APPEND PROTOC_INC_DIRS "-I${FILE_PATH}")
+    list(APPEND PROTO_BYPRODUCTS "${PROTO_COMPILATION_DIR}/${FILE_BASE}.pb.c")
+endforeach()
+
+# Populate list of compilation commands for protoc compiler.
+foreach(file ${PROTO_FILES})
+    get_filename_component(FILE_PATH ${file} DIRECTORY)
+    get_filename_component(FILE_NAME ${file} NAME)
+    set(PROTOC_ARGS "${PROTOC_INC_DIRS}")
+    list(APPEND PROTOC_ARGS "--nanopb_out=${PROTO_COMPILATION_DIR}")
+    list(APPEND PROTOC_ARGS "${FILE_NAME}")
+    list(APPEND PROTOC_CMD_LIST COMMAND ${CMAKE_COMMAND} -E chdir ${FILE_PATH} ${PROTOC_BIN} ${PROTOC_ARGS})
+endforeach()
+
+# Add custom target to compile all .proto files
+add_custom_target(
+    protobuf_generation
+    COMMAND ${PROTOC_CMD_LIST}
+    BYPRODUCTS ${PROTO_BYPRODUCTS}
+)
+
+zephyr_library_named(ALEXA_GADGETS_PROFILE)
+
+zephyr_include_directories(${CMAKE_CURRENT_SOURCE_DIR})
+zephyr_include_directories(${PROTO_COMPILATION_DIR})
+zephyr_include_directories(${NANOPB_PATH})
+
+zephyr_library_sources_ifdef(CONFIG_GADGETS_PROFILE_ENABLE
+    ${CMAKE_CURRENT_SOURCE_DIR}/gadgets_profile.c
+    ${NANOPB_PATH}/pb_common.c
+    ${NANOPB_PATH}/pb_encode.c
+    ${NANOPB_PATH}/pb_decode.c
+    ${PROTO_COMPILATION_DIR}/accessories.pb.c
+    ${PROTO_COMPILATION_DIR}/common.pb.c
+    ${PROTO_COMPILATION_DIR}/device.pb.c
+    ${PROTO_COMPILATION_DIR}/firmware.pb.c
+    ${PROTO_COMPILATION_DIR}/directiveParser.pb.c
+    ${PROTO_COMPILATION_DIR}/directiveHeader.pb.c
+    ${PROTO_COMPILATION_DIR}/eventParser.pb.c
+    ${PROTO_COMPILATION_DIR}/eventHeader.pb.c
+    ${PROTO_COMPILATION_DIR}/alexaDiscoveryDiscoverDirective.pb.c
+    ${PROTO_COMPILATION_DIR}/alexaDiscoveryDiscoverResponseEvent.pb.c
+    ${PROTO_COMPILATION_DIR}/alexaDiscoveryDiscoverResponseEventPayload.pb.c)
```

Comment:
If you make a couple of lists above, then you could use the `PROTO_BYPRODUCTS` list instead, and thus have just a single location where you describe the files.

As example:
```
set(PROFILE_PROTO_FILES
      <files>.proto
)

set(NOTIFICATIONS_PROTO_FILES
      <files>.proto
)

..... continue the pattern ...

foreach(GROUP PROFILE NOTIFICATIONS ....)
    # Populate list of expected .c files output from protobuf generation
    foreach(file ${${GROUP}_PROTO_FILES})
         ... same as now ...
         list(APPEND ${GROUP}_PROTO_BYPRODUCTS "${PROTO_COMPILATION_DIR}/${FILE_BASE}.pb.c")
    endforeach()

    # Populate list of compilation commands for protoc compiler.
    foreach(file ${${GROUP}_PROTO_FILES})
         ... same as now ...
    endforeach()

    add_custom_target(
        protobuf_generation_${GROUP}
        COMMAND ${PROTOC_CMD_LIST}
        BYPRODUCTS ${${GROUP}_PROTO_BYPRODUCTS}
    )
    endforeach()
endforeach()

....

zephyr_library_sources_ifdef(CONFIG_GADGETS_PROFILE_ENABLE
    ${CMAKE_CURRENT_SOURCE_DIR}/gadgets_profile.c
    ${NANOPB_PATH}/pb_common.c
    ${NANOPB_PATH}/pb_encode.c
    ${NANOPB_PATH}/pb_decode.c
    ${PROFILE_PROTO_BYPRODUCTS}
)

 zephyr_library_sources_ifdef(CONFIG_GADGETS_CAPABILITY_NOTIFICATIONS_ENABLE
    ${NOTIFICATIONS_PROTO_BYPRODUCTS}
)

..... continue the pattern ...

```

This further has the advantages that only the required proto files are compiled into C-files.
So if `CONFIG_GADGETS_CAPABILITY_NOTIFICATIONS_ENABLE=n` then `protobuf_generation_NOTIFICATIONS` are not needed, and not compiled.

Currently, all `.proto` files are always compiled into C-files, even if the C-file is not needed because the library is disabled.

Reply:
I like the idea of separating the proto files into groups like this. Note that a consequence of this organization is that all the proto-generated header files needs #ifdefs in case this particular capability is disabled.

Reply:
Does this mean that `headerA` includes `headerB`, and `headerA` is included by `sourceA` ?
So that one still requires `headerB` even if `featureB` is disabled.

How does that work now, in case `featureB` is disabled, cause then the functions defined in `headerB` cannot be linked.
Are there `ifdef`'s in the source code to ensure this won't break ?

You may still do the split, but keeping a single custom target, similar to:
```
    endforeach()
    list(APPEND PROTO_BYPRODUCTS ${${GROUP}_PROTO_BYPRODUCTS})
endforeach()

add_custom_target(
      protobuf_generation
      COMMAND ${PROTOC_CMD_LIST}
      BYPRODUCTS ${PROTO_BYPRODUCTS}
)
```
that way you'll have a single target, but still individual sublists to use in `zephyr_library_sources()`

Reply:
The dependencies are luckily not that complicated. Feature-level headers have dependencies to the "PROFILE" group, but not vice versa.

The ifdefs I was referring to would be in gadgets_profile.c, which uses the generated headers. 
E.g. for the "music data" feature:
```
#if CONFIG_GADGETS_CAPABILITY_MUSICDATA_ENABLE
#include "alexaGadgetMusicDataTempoDirective.pb.h"
#include "alexaGadgetMusicDataTempoDirectivePayload.pb.h"
#endif
```

Reply:
ok. then the ifdef sounds like a good approach.

For a second I was concerned you talked about ifdefs in some generated headers.